pkg/controlplane/clientutils/multiclusterconfig.go

```diff
@@ -0,0 +1,147 @@
+	"k8s.io/klog"
+)
+
+
+type multiClusterClientConfigRoundTripper struct {
+	rt            http.RoundTripper
+	requestInfoResolver func() genericapirequest.RequestInfoResolver
+	enabledOn     sets.String
+}
+
+func EnableMultiCluster(config *rest.Config, apiServerConfig *genericapiserver.Config, enabledOnResources... string) {
+	config.ContentConfig.ContentType = "application/json"
+	
+	config.Wrap(func(rt http.RoundTripper) http.RoundTripper {
+		defaultResolver := defaultRequestInfoResolver()
+		return &multiClusterClientConfigRoundTripper{
+			rt:            rt,
+			requestInfoResolver: func() genericapirequest.RequestInfoResolver {
+				if apiServerConfig != nil {
+					return apiServerConfig.RequestInfoResolver
+				} else {
+					return defaultResolver
+				}
+			},
+			enabledOn: sets.NewString(enabledOnResources...),
+		}
+	})
+}
+
+func defaultRequestInfoResolver() genericapirequest.RequestInfoResolver {
+	apiPrefixes := sets.NewString(strings.Trim(server.APIGroupPrefix, "/")) // all possible API prefixes
+	legacyAPIPrefixes := sets.String{}                               // APIPrefixes that won't have groups (legacy)
+	apiPrefixes.Insert(strings.Trim(server.DefaultLegacyAPIPrefix, "/"))
+	legacyAPIPrefixes.Insert(strings.Trim(server.DefaultLegacyAPIPrefix, "/"))
+
+	return &request.RequestInfoFactory{
+		APIPrefixes:          apiPrefixes,
+		GrouplessAPIPrefixes: legacyAPIPrefixes,
+	}
+}
+
+func (mcrt *multiClusterClientConfigRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	req = utilnet.CloneRequest(req)
+	requestInfo, err := mcrt.requestInfoResolver().NewRequestInfo(req)
+	if err != nil {
+		return nil, err
+	}
+	if requestInfo != nil &&
+		mcrt.enabledOn.Has(requestInfo.Resource) {
+		contextCluster := genericapirequest.ClusterFrom(req.Context())
+		resourceClusterName := ""
+		headerCluster := ""
+		switch requestInfo.Verb {
+		case "list", "watch":
+			if contextCluster != nil && ! contextCluster.Wildcard {
+				headerCluster = contextCluster.Name
+			} else {
+				headerCluster = "*"
+			}
+		case "create", "update":
+			err := func() error {
+				reader, err := req.GetBody()
+				if err != nil {
+					klog.Infof("DEBUG: Error when trying to read the request body from the multicluster client config roundtripper: %v", err)
+					return err
+				}
+				defer reader.Close()
+				buf := new(bytes.Buffer)
+				_, err = buf.ReadFrom(reader)
+				if err != nil {
+					klog.Infof("DEBUG: Error when trying to read the request body from the multicluster client config roundtripper: %v", err)
+					return err
+				}
+				bytes := buf.Bytes()
+				obj, _, err := unstructured.UnstructuredJSONScheme.Decode(bytes, nil, nil)
+				if err != nil {
+					klog.Infof("DEBUG: Error when trying to read the request body from the multicluster client config roundtripper: %v", err)
+					return err
+				}
+				if s, ok := obj.(metav1.Object); ok {
+					resourceClusterName = s.GetClusterName()
```

Comment:
Yeah... you can't do this... :)  You need a wrapper at a much higher level, because this breaks most of the client stack.  We'll have to talk about where you could actually do this at (it'll be something in client-go).

Reply:
Roughly we have to get to the object before it gets serialized, not after

Reply:
Maybe a comment is required here. There are a number of cases (typically in controllers) when you list or watch objects with wildcards, and in the resulting objects you have the `ClusterName`  field set.
Then you want to update this object (its status for example): It should be updated in the right cluster. That's the reason of this code here. Set the cluster of the client request according to the `ClusterName` field if it is already filled (probably by a previous List or Watch call).

Reply:
For posterity, in a follow up to this we'll put the proper hooks in client-go/rest for this.